File: LeetCode/py/811.py
#Link: https://leetcode.com/problems/subdomain-visit-count/

class Solution:
    subdomains = {}
    def subdomainVisits(self, cpdomains: List[str]) -> List[str]:
        self.subdomains = {}
        for domain in cpdomains:
            count = int(domain.split()[0])
            domain = domain.split()[1]
            self.addSubdomain(domain, count)
            subdomain = ""
            for ch in domain[::-1]:
                if ch == '.': self.addSubdomain(subdomain, count)
                subdomain = ch + subdomain
        return self.getResult()
    def addSubdomain(self, new: str, count: int):
        if new in self.subdomains: self.subdomains[new] += count
        else: self.subdomains[new] = count
    def getResult(self) -> List[str]:
        temp = []
        for x in self.subdomains:
            temp.append(str(self.subdomains[x]) + " " + x)
        return temp

File: LeetCode/cpp/811.cpp
//Link: https://leetcode.com/problems/subdomain-visit-count/

class Solution {
private:
    vector<pair<string,int>> subdomains;
    void add(string subdomain, int count) {
        for(int i = 0 ; i < subdomains.size() ; i++) {
            if(subdomains[i].first == subdomain) {
                subdomains[i].second += count;
                return ;
            }
        }
        subdomains.push_back(make_pair(subdomain,count));
    }
public:
    vector<string> subdomainVisits(vector<string>& cpdomains) {
        vector<string> result;
        int count;
        string subdomain;
        for(auto domain : cpdomains) {
            count = 0;
            subdomain = "";
            for(int i = 0 ; domain[i] != ' ' ; i++)
                count = count*10 + domain[i]-48;
            for(int i = domain.length()-1 ; i >= 0 ; i--) {
                if(domain[i] == '.' || domain[i] == ' ') add(subdomain, count);
                subdomain = domain[i] + subdomain;
            }
        }
        for(auto x : subdomains)
            result.push_back(to_string(x.second)+" "+x.first);
        return result;
    }
};
